unblob/file_utils.py

```diff
@@ -332,3 +353,147 @@ def read_until_past(file: File, pattern: bytes):
             return file.tell()
         if next_byte not in pattern:
             return file.tell() - 1
+
+
+class Sandbox:
+    """Restricts file system operations to a directory.
+
+    All operations succeed silently (without exceptions), but errors
+    (e.g. path traversal violations) are collected as a list of :ExtractionProblem:-s.
+
+    All input paths are expected to be absolute paths (except for link targets).
+
+    This class is not the place of smartness, all path transformations,
+    like converting relative paths or absolute paths to paths under the extraction root
+    remains the responsibility of the user.
+    """
+
+    problems: List[ExtractionProblem]
+
+    def __init__(self, root: Path):
+        self.root = root.resolve()
+        self.problems = []
+
+    def record_problem(self, problem: ExtractionProblem):
+        self.problems.append(problem)
+
+        msg = f"{problem.problem} {problem.resolution}"
+        if isinstance(problem, LinkExtractionProblem):
+            logger.warning(msg, path=problem.path, link_path=problem.link_path)
+        else:
+            logger.warning(msg, path=problem.path)
```

Comment:
`ExtractionProblem` should grow a `.log_me_with(logger)` method to avoid this `isinstance()` and support potential new detailed subclasses of `ExtractionProblem`.